Limit changes and config validation

Heufneutje opened the pull request:
Since we're still in the breakage phase of txircd 0.4, I give you a breaking, yet valuable in my opinion, change.

So let's go over the things that are in this PR:
- Replacement of a lot of magic numbers with configurable equivalents. I kinda took this upon myself because the amount of magic numbers we have bothered me. Making these configurable is, in my opinion, a change that will not only make the codebase cleaner and assure that these numbers are always the same in all cases (see my fix in TOPICLEN for an example on this one), but also make txircd a lot more configurable for our users. One could argue that a lot of these numbers should never be changed, but I prefer going the InspIRCd route of "configurable but discouraged" over "not configurable at all". Besides, we already had some of them configurable. It doesn't really make sense to make some configurable and some hardcoded.

- The addition of a few new limits. We did not yet have limits on our gecos, our AWAY messages and our KICK reasons. In my opinion they should be present since I was working on the limits anyway, I added them. The default values were taken from InspIRCd.

- Modularization of extra ISUPPORT tokens. The last change is for the ISUPPORT dictionary generation, which now allows modules to add their own tokens on top of the default ones we already have. In theory we could even move the tokens that belong to core modules to their respective modules, but I'll let you be the judge on that one.

ElementalAlchemist commented:
I like what we're doing here, but there are a few things here that we should do for this to be a good thing to include.

1. We should enforce a maximum cap on all of the limits. Many of the "magic numbers" (which were not magic at all, actually) were chosen because they would cap the line at exactly 512 characters (assuming all limits were maxed, e.g. with a 32-char nick, 12-char ident, and 64-char host in the prefix, for example, and a 64-char channel name where that's also in the message). (A couple of them, like the 255-char max on mode parameters, are a bit more arbitrary, but they're all intended to keep messages to a reasonable size.) If we do this, we should actively prevent admins from configuring their IRCd to make messages longer than 512 chars by enforcing a cap on the lengths. Most of the caps should be whatever the cap was before we made it configurable.
2. I don't know what you mean by "modes_length". Try something like "modes_per_line" or "mode_changes_per_line".
3. You've done something really bad, which is to violate the encapsulation going on here. The core should be making no assumptions about loaded modules (even if they're core modules)--but here we're assuming everything about ISUPPORT. If we're going to do this, let's have an action (it'll be standard) to let modules add to the ISUPPORT dict.

Heufneutje commented:
1. I know why the limits were configured the way they were. Think we should handle this in config parsing? I think sanitizing a lot of the config values there wouldn't be a bad thing honestly.
2. Will do.
3. Sounds good.

ElementalAlchemist commented:
1. Yeah, which is fine, but I'm saying we should make it idiot-proof. Probably the best way to do it is with some sort of `config.getNumberWithMaximum` function or something, which takes a key, a maximum value, and a default value.

Heufneutje commented:
In my opinion the entire config could use some idiot-proofing. Right now it's very easy to insert garbage since Python's typing is so dynamic. I've thought about creating a `config.getNumberWithMax` but came to the same conclusion that we should yell at the admin for configuring a value that's too high.

Heufneutje commented:
I also went ahead and fixed the modes config option. I removed the ISUPPORT change from this PR so this is now just about the limit changes I'll change ISUPPORT properly in another PR (I'd like to talk to you about what you have in mind for it before I do).

Heufneutje commented:
This is ready now. Let me know if you find anything that should be changed.

ElementalAlchemist commented on txircd/ircd.py:
```diff
@@ -170,6 +166,12 @@ def _loadModuleData(self, module):
 		common = False
+
+		try:
+			module.verifyConfig(self.config)
```
We should do this way earlier (maybe right after `hookIRCd`).

ElementalAlchemist commented on txircd/ircd.py:
```diff
@@ -372,7 +374,111 @@ def reloadModule(self, moduleName):
+		if not isinstance(config["bind_client"], list):
+			raise ConfigValidationError("bind_client", "value must be a list")
+		for bindDesc in config["bind_client"]:
+			if not isinstance(bindDesc, str):
```
In py2, the appropriate way to check for a string is with `isinstance(stringToCheck, basestring)`.

ElementalAlchemist commented on txircd/ircd.py:
```diff
@@ -372,7 +374,111 @@ def reloadModule(self, moduleName):
+				if not isinstance(desc, str):
+					raise ConfigValidationError("links", "\"{}\" is an invalid server description".format(desc))
+				if not isinstance(server, dict):
+					raise ConfigValidationError("links", "values for \"{}\" must be a dictionary".format(desc))
```
We're verifying all the other types; are we not going to validate the keys/values of this dictionary, too? :P  (This includes all the other dictionary checks, too.)

ElementalAlchemist commented on txircd/ircd.py:
```diff
@@ -372,7 +374,111 @@ def reloadModule(self, moduleName):
+			raise ConfigValidationError("server_ping_frequency", "invalid number")
+
+		for module in self.loadedModules.itervalues():
+			module.veryifyConfig(config)
```
This still says `veryifyConfig`.

ElementalAlchemist commented on txircd/ircd.py:
```diff
@@ -449,10 +555,11 @@ def generateISupportList(self):
 		statusSymbolOrder = "".join([self.channelStatuses[status][0] for status in self.channelStatusOrder])
 		isupport["CHANMODES"] = ",".join(["".join(modes) for modes in self.channelModes])
+		isupport["CHANNELEN"] = self.config.get("channel_name_length", 64)
+		isupport["NETWORK"] = self.config["network_name"]
```
You removed a bunch of hardcoded values above, but I don't see them here, nor do I see any actions to pull them in.

Heufneutje replied:
The action has been added now.

ElementalAlchemist commented on txircd/modules/extra/cloaking.py:
```diff
@@ -21,6 +22,22 @@ def userModes(self):
+	def load(self):
+		self.rehash()
+
+	def rehash(self):
```
Is there any value anymore in having this function?

Heufneutje replied:
Not anymore now.

ElementalAlchemist commented on txircd/modules/extra/ratelimit.py:
```diff
@@ -11,6 +12,16 @@ class RateLimit(ModuleData):
+			if not isinstance(config["ratelimit"], dict):
+				raise ConfigValidationError("ratelimit", "value must be a dictionary")
+			for key, value in config["ratelimit"].itemitems():
+				if key not in ["limit", "kill_limit", "interval"]:
```
If you use parentheses instead, it'll make it a tuple.
I don't know if there's a difference that matters for this purpose, but tuples are what we use everywhere else we do this.

ElementalAlchemist commented on txircd/modules/rfc/cmd_away.py:
```diff
@@ -16,6 +17,20 @@ def actions(self):
+		self.rehash()
+
+	def rehash(self):
+		self.ircd.isupport_tokens["AWAYLEN"] = self.ircd.config.get("away_length", 200)
```
If we're going to do THIS, we need to document somewhere for non-core modules that it MUST be removed on unload.

ElementalAlchemist commented on txircd/modules/rfc/cmd_away.py:
```diff
@@ -16,6 +17,20 @@ def actions(self):
+			if not isinstance(config["away_length"], int) or config["away_length"] < 0:
+				raise ConfigValidationError("away_length", "invalid number")
+			elif config["away_length"] > 200:
+				config["away_length"] = 200
```
This won't work on load since `__setitem__` isn't defined for Config. (This will just raise an error.)

Heufneutje replied:
Has been fixed.

ElementalAlchemist commented on txircd/modules/server/autoconnect.py:
```diff
@@ -25,7 +26,17 @@ def rehash(self):
+
+	def verifyConfig(self, config):
+		if "autoconnect_period" in config and (not isinstance(config["autoconnect_period"], int) or config["autoconnect_period"] < 0):
+				raise ConfigValidationError("autoconnect_period", "invalid number")
```
Why is this indented twice?

Heufneutje replied:
Because oops.

ElementalAlchemist commented on txircd/ircd.py:
```diff
@@ -372,7 +374,111 @@ def reloadModule(self, moduleName):
+		for module in self.loadedModules.itervalues():
+			module.veryifyConfig(config)
+
+	def logConfigValidationWarning(self, key, message, default):
```
Does this need to be in IRCd? Can it be in utils?

Heufneutje replied:
It can.

Heufneutje replied:
With the new logging system I've kept it in IRCd as we discussed on IRC.

ElementalAlchemist commented:
I left a hundred inline comments but still have one other thing:
We should also take the config checks from Config (required, requiredValue, formatValue) and move them to the new validation infrastructure so that it's all in one place.

ElementalAlchemist commented on txircd/ircd.py:
```diff
@@ -385,7 +375,150 @@ def reloadModule(self, moduleName):
-	
+
+	def verifyConfig(self, config):
+		# Verify the configuration values that are used in the core first.
```
These comments give me no additional information.

ElementalAlchemist commented on txircd/ircd.py:
```diff
@@ -385,7 +375,150 @@ def reloadModule(self, moduleName):
+			raise ConfigValidationError("server_name", "value must be a string")
+		if len(config["server_name"]) > 64:
+			config["server_name"] = config["server_name"][:64]
+			self.logConfigValidationWarning("server_name", "value is too long", config["server_name"])
```
Can we add to the warning that the server name was truncated?

ElementalAlchemist commented on txircd/ircd.py:
```diff
@@ -385,7 +375,150 @@ def reloadModule(self, moduleName):
+			randFromName = random.Random(config["server_name"])
+			config["server_id"] = randFromName.choice(string.digits)
+			config["server_id"] += randFromName.choice(string.digits + string.uppercase)
+			config["server_id"] += randFromName.choice(string.digits + string.uppercase)
```
Does this need to be three lines?
If yes, can we assign a local string and assign that to the dict at the end?

ElementalAlchemist commented on txircd/ircd.py:
```diff
@@ -385,7 +375,150 @@ def reloadModule(self, moduleName):
+		if "server_description" not in config:
+			raise ConfigValidationError("server_description", "required item not found in configuration file.")
+		if not isinstance(config["server_description"], basestring):
+			raise ConfigValidationError("server_description", "value must be a string")
```
We should also check that server_description's value is not the empty string.

Heufneutje replied:
Done.

ElementalAlchemist commented on txircd/ircd.py:
```diff
@@ -385,7 +375,150 @@ def reloadModule(self, moduleName):
+			raise ConfigValidationError("network_name", "value must be a string")
+		if len(config["network_name"]) > 32:
+			config["network_name"] = config["network_name"][:32]
+			self.logConfigValidationWarning("network_name", "value is too long", config["network_name"])
```
We should also check that network_name's value is not the empty string.

ElementalAlchemist commented on txircd/ircd.py:
```diff
@@ -385,7 +375,150 @@ def reloadModule(self, moduleName):
+			self.logConfigValidationWarning("network_name", "value is too long", config["network_name"])
+		if not "bind_client" not in config:
+			config["bind_client"] = [ "tcp:6667:interface={::}" ]
+			self.logConfigValidationWarning("bind_client", "no default client binding specified", "[ \"tcp:6667:interface={::}\" ]")
```
Should this be a warning? I'd put it at LogLevel.info.

Heufneutje replied:
I think it should be a warning. In my opinion this default shouldn't exist in the first place or be bound to localhost only. But that's just me.

ElementalAlchemist commented on txircd/ircd.py:
```diff
@@ -385,7 +375,150 @@ def reloadModule(self, moduleName):
+				if not isinstance(server, dict):
+					raise ConfigValidationError("links", "values for \"{}\" must be a dictionary".format(desc))
+				if "connect_descriptor" not in server:
+					raise ConfigValidationError("links", "server \"{}\" must contain a \"connect_descriptor\" value".format(desc))
```
If we're checking this here, we should be able to remove the check from `connectServer`.

Heufneutje replied:
Done.

ElementalAlchemist commented on txircd/ircd.py:
```diff
@@ -385,7 +375,150 @@ def reloadModule(self, moduleName):
+					if not isinstance(server["out_password"], basestring):
+						config["links"][desc]["out_password"] = str(server["out_password"])
+		if "datastore_path" not in config:
+			config["datastore_path"] = "data.db"
```
We can also remove the `.get` call for this if we're just going to default it here.

Heufneutje replied:
Done. Will fix the OS check tomorrow.

Heufneutje replied:
And fixed.

ElementalAlchemist commented on txircd/ircd.py:
```diff
@@ -385,7 +375,150 @@ def reloadModule(self, moduleName):
+			if not isinstance(config["channel_name_length"], int) or config["channel_name_length"] < 0:
+				raise ConfigValidationError("channel_name_length", "invalid number")
+			elif config["channel_name_length"] > 32:
+				config["channel_name_length"] = 32
```
This is 64.

ElementalAlchemist commented on txircd/ircd.py:
```diff
@@ -385,7 +375,150 @@ def reloadModule(self, moduleName):
+				config["modes_per_line"] = 20
+				self.logConfigValidationWarning("modes_per_line", "value is too large", 20)
+		if "channel_listmode_limit" in config and (not isinstance(config["channel_listmode_limit"], int) or config["channel_listmode_limit"] < 0):
+			raise ConfigValidationError("channel_listmode_limit", "invalid number")
```
Should there be a cap on this one, too?  Maybe 256 or 512?

Heufneutje replied:
We could. Would anyone actually be crazy enough to set 512 listmodes though?

ElementalAlchemist replied:
Maybe not; I wouldn't put it above 512. Even if it's just 256, that's fine.

Heufneutje replied:
256 it will be then.

ElementalAlchemist commented on txircd/ircd.py:
```diff
@@ -470,10 +603,16 @@ def generateISupportList(self):
+
+		extraTokens = {}
+		self.runActionStandard("buildisupport", extraTokens)
+		isupport.update(extraTokens)
```
If we're going to do this, why make a separate dict? We should
- pass the `isupport` dict if we don't care about the existing values
- if we want to enforce safety, use this dict to update the isupport dict in a way that doesn't overwrite preexisting values.

ElementalAlchemist commented on txircd/module_interface.py:
```diff
@@ -108,6 +108,13 @@ def fullUnload():
+		"""
+		Called when the module is loaded and when the server is rehashed.  Should check all
+		configuration values defined by this module and either adjust them in the configuration
+		or raise a ConfigError when an invalid value is defined.
```
`ConfigValidationError` is probably the only `ConfigError` that should be coming out of these.

ElementalAlchemist commented on txircd/modules/extra/cloaking.py:
```diff
@@ -19,6 +20,15 @@ def userModes(self):
+			if not isinstance(config["cloaking_salt"], basestring):
+				raise ConfigValidationError("cloaking_salt", "value must be a string")
+		else:
+			self.ircd.log.warn("No cloaking salt was found in the config. Host cloaks will be insecure!")
```
We should probably log this warning even if the value is specified but empty.

ElementalAlchemist commented on txircd/modules/extra/customprefix.py:
```diff
@@ -13,15 +14,25 @@ def channelModes(self):
+				raise ConfigValidationError("custom_prefixes", "value must be a dictionary")
+			for prefix, prefixValue in config["custom_prefixes"]:
+				if len(prefix) > 1:
+					raise ConfigValidationError("custom_prefixes", "prefix value \"{}\" should be a mode character")
```
What if it's less than 1 or not a string?

ElementalAlchemist commented on txircd/modules/extra/customprefix.py:
```diff
@@ -13,15 +14,25 @@ def channelModes(self):
+					raise ConfigValidationError("custom_prefixes", "value \"char\" for prefix \"{}\" is missing".format(prefix))
+				if not isinstance(prefixValue["level"], int):
+					raise ConfigValidationError("custom_prefixes", "prefix \"{}\" does not specify a valid level")
+				if not isinstance(prefixValue["char"], basestring) or len(prefixValue["char"]) > 1:
```
What if it's an empty string?

ElementalAlchemist commented on txircd/modules/extra/denychans.py:
```diff
@@ -13,6 +15,17 @@ class DenyChannels(ModuleData):
+				for chanName in config[option]:
+					if chanName[0] != "#":
+						chanName = "#{}".format(chanName)
+					if not isValidChannelName(chanName):
```
This breaks wildcards for channel names. It's probably easier not to check this.

ElementalAlchemist commented on txircd/modules/rfc/cmd_nick.py:
```diff
@@ -14,13 +15,22 @@ class NickCommand(ModuleData):
 				("changenick", 1, self.broadcastNickChange),
-				("remotechangenick", 1, self.broadcastNickChange) ]
+				("remotechangenick", 1, self.broadcastNickChange),
+		        ("buildisupport", 1, self.buildISupport) ]
```
The indenting is weird for this entire list for some reason.

ElementalAlchemist commented on txircd/modules/rfc/mech_oper.py:
```diff
@@ -24,6 +25,32 @@ def actions(self):
+					raise ConfigValidationError("opers", "no password defined for oper \"{}\"".format(operNick))
+				if not isinstance(values["password"], basestring):
+					raise ConfigValidationError("opers", "value \"password\" for oper \"{}\" must be string".format(operNick))
+				if "hash" in values and values["hash"] != "pbkdf2": # Change later if we add more hashing modules
```
Can we not make assumptions about other modules here?'
We should not make assumptions about other modules (especially non-core modules like those in stringhash).

ElementalAlchemist commented on txircd/modules/rfc/mech_oper.py:
```diff
@@ -24,6 +25,32 @@ def actions(self):
+						if not isinstance(operType, basestring):
+							raise ConfigValidationError("opers", "every type entry for oper \"{}\" must be string".format(operNick))
+		if "oper_types" in config:
+			if not isinstance(config["oper_types"], list):
```
oper_types is a dict.

ElementalAlchemist commented on txircd/ircd.py:
```diff
@@ -6,16 +6,16 @@
 from txircd.utils import CaseInsensitiveDictionary, ModeType, now, unescapeEndpointDescription
 from weakref import WeakValueDictionary
-import importlib, random, shelve, string, txircd.modules
+import importlib, random, re, os, shelve, string, txircd.modules
```
`os` wasn't added in even a little bit of alphabetical order.

ElementalAlchemist commented on txircd/modules/rfc/mech_oper.py:
```diff
@@ -24,6 +25,32 @@ def actions(self):
+		if "oper_types" in config:
+			if not isinstance(config["oper_types"], dict):
+				raise ConfigValidationError("oper_types", "value must be a dictionary")
+			for permission in config["oper_types"].itervalues():
```
Did you not want to check that all the keys are strings and that the value is a list before you iterate over it?
